File: Different/mergeSortedArraysWithoutDuplicates.js
const { test } = require('./test');

// Даны два массива целых чисел. Массивы отсортированы по возрастанию. Вывести объединение без дублей, сохранив сортировку и использовав константу дополнительной памяти.
const mergeSortedArraysWithoutDuplicates = (arr1, arr2) => {
  const result = [];

  const length1 = arr1.length;
  const length2 = arr2.length;

  let first = 0;
  let second = 0;

  let last = null;

  while (first < length1 || second < length2) {
    if (second >= length2 || (first < length1 && arr1[first] < arr2[second])) {
      if (last === null || arr1[first] > last) {
        result.push(arr1[first]);
        last = arr1[first];
      }

      first++;

      continue;
    }

    if (last === null || arr2[second] > last) {
      result.push(arr2[second]);
      last = arr2[second];
    }

    second++;
  }

  return result.join('-');
};

test(mergeSortedArraysWithoutDuplicates, [
  {
    input: [
      [1, 3, 5, 7, 9],
      [2, 4, 6, 8],
    ],
    expected: [1, 2, 3, 4, 5, 6, 7, 8, 9].join('-'),
  },
  {
    input: [
      [2, 4, 6],
      [4, 6, 8],
    ],
    expected: [2, 4, 6, 8].join('-'),
  },
  {
    input: [[], [1, 2, 3]],
    expected: [1, 2, 3].join('-'),
  },
  {
    input: [[], []],
    expected: [].join('-'),
  },
  {
    input: [
      [1, 2, 3, 4, 5],
      [2, 3, 4],
    ],
    expected: [1, 2, 3, 4, 5].join('-'),
  },
  {
    input: [
      [2, 2, 2],
      [2, 2],
    ],
    expected: [2].join('-'),
  },
  {
    input: [
      [1, 2, 3, 4, 5, 6, 7],
      [2, 4, 6],
    ],
    expected: [1, 2, 3, 4, 5, 6, 7].join('-'),
  },
  {
    input: [
      [1, 2, 3],
      [3, 4, 5],
    ],
    expected: [1, 2, 3, 4, 5].join('-'),
  },
  {
    input: [
      [-5, -3, -1],
      [-4, -3, -2, -1, 0],
    ],
    expected: [-5, -4, -3, -2, -1, 0].join('-'),
  },
  {
    input: [
      [1, 1, 2, 2, 3],
      [2, 2, 3, 3, 4],
    ],
    expected: [1, 2, 3, 4].join('-'),
  },
  {
    input: [
      [1, 1, 1],
      [1, 1, 1, 1],
    ],
    expected: [1].join('-'),
  },
]);

File: Different/mergeSortedArraysWithoutDuplicates.py
# Даны два массива целых чисел. Массивы отсортированы по возрастанию. Вывести объединение без дублей, сохранив сортировку и использовав константу дополнительной памяти.

from sys import maxsize


class Solution:
    def maxSumSubarrayWithTwoUnique(self, nums: list[int]) -> int | str:
        if len(nums) == 0:
            return "Empty list"

        max_sum = -maxsize - 1

        right = 0
        left = 0

        my_dict: dict[int, int] = {}

        current_sum = 0

        queue: list[int] = []

        while right < len(nums):
            current = nums[right]

            if current not in my_dict:
                queue.append(current)

            if len(queue) == 2 and current == queue[0]:
                queue.pop(0)
                queue.append(current)

            if len(queue) > 2:
                num_to_delete = queue.pop(0)

                index = my_dict[num_to_delete]

                while left <= index:
                    current_sum -= nums[left]
                    left += 1

                del my_dict[num_to_delete]

            my_dict[current] = right

            current_sum += current
            max_sum = max(max_sum, current_sum)

            right += 1

        return max_sum


solution = Solution()


def test(callback, cases):
    for case in cases:
        input_value = case["input"]
        expected_value = case["expected"]

        result = callback(*input_value)

        print("input", input_value)
        print("result", result)
        print("expected", expected_value)

        print(
            "\x1b[32mPassed\x1b[0m"
            if expected_value == result
            else "\x1b[31mFailed\x1b[0m"
        )

        print("-----------------------------------")


test(
    solution.maxSumSubarrayWithTwoUnique,
    [
        {
            "input": [[1, 2, 3, 2, 2]],
            "expected": 9,
        },
        {
            "input": [[5, 5, 5, 5]],
            "expected": 20,
        },
        {
            "input": [[3, 3, 2, 2, 3]],
            "expected": 13,
        },
        {
            "input": [[1, 2, 3, 4, 5]],
            "expected": 9,
        },
        {
            "input": [[10]],
            "expected": 10,
        },
        {
            "input": [[2, 2, 2]],
            "expected": 6,
        },
        {
            "input": [[-1, -2, -3, -2, -2]],
            "expected": -1,
        },
        {
            "input": [[1, 2, 1, 3, 4, 3, 5, 3, 3, 2, 1]],
            "expected": 14,
        },
    ],
)
